// maxun-sdk/dist/builders/extract-builder.js
"use strict";
/**
 * Extract-specific workflow builder
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtractBuilder = void 0;
const workflow_builder_1 = require("./workflow-builder");
class ExtractBuilder extends workflow_builder_1.WorkflowBuilder {
    constructor(name) {
        super(name, 'extract');
    }
    /**
     * Set the parent extractor (called internally)
     */
    setExtractor(extractor) {
        this.extractor = extractor;
        return this;
    }
    /**
     * Capture specific text fields from the page
     */
    captureText(fields, name) {
        // Pass fields directly as plain selectors
        this.addAction({
            action: 'scrapeSchema',
            args: [fields],
            name: name,
        });
        return this;
    }
    /**
     * Capture a list of items with pagination support
     * All fields are automatically detected and extracted
     *
     * @param config - List extraction configuration
     * @param config.selector - CSS selector for list items
     * @param config.pagination - Optional pagination configuration
     * @param config.maxItems - Maximum number of items to extract (default: 100)
     * @param name - Optional action name
     */
    captureList(config, name) {
        const scrapeListConfig = {
            itemSelector: config.selector,
            maxItems: config.maxItems || 100
        };
        if (config.pagination) {
            scrapeListConfig.pagination = {
                type: config.pagination.type,
                selector: config.pagination.selector || null
            };
        }
        this.addAction({
            action: 'scrapeList',
            args: [scrapeListConfig],
            name: name,
        });
        return this;
    }
    /**
     * Make the builder awaitable - converts builder to Robot instance
     */
    then(onfulfilled, onrejected) {
        if (!this.extractor) {
            return Promise.reject(new Error('Builder not properly initialized. Use extractor.create() to create a builder.')).then(onfulfilled, onrejected);
        }
        // Build and convert to Robot
        return this.extractor.build(this).then(onfulfilled, onrejected);
    }
}
exports.ExtractBuilder = ExtractBuilder;
