// maxun-sdk/dist/builders/workflow-builder.js
"use strict";
/**
 * Base Workflow Builder
 * Provides common functionality for building Maxun workflows
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowBuilder = void 0;
class WorkflowBuilder {
    constructor(name, robotType) {
        this.name = name;
        this.robotType = robotType;
        this.workflow = [];
        this.meta = {};
        this.currentStep = null;
        this.isFirstNavigation = true;
        this.meta.name = name;
        this.meta.robotType = robotType;
    }
    /**
     * Navigate to a URL
     */
    navigate(url) {
        // Create the main step (no goto action here)
        const mainStep = {
            where: { url },
            what: [],
        };
        // Only add about:blank on FIRST navigation
        if (this.isFirstNavigation) {
            // Create the about:blank step
            const aboutBlankStep = {
                where: { url: 'about:blank' },
                what: [
                    {
                        action: 'goto',
                        args: [url],
                    },
                    {
                        action: 'waitForLoadState',
                        args: ['networkidle'],
                    },
                ],
            };
            // Add main step to front, about:blank to end (bottom)
            this.workflow.unshift(mainStep);
            this.workflow.push(aboutBlankStep);
            this.isFirstNavigation = false;
        }
        else {
            // Subsequent navigations: add to front
            this.workflow.unshift(mainStep);
        }
        this.currentStep = mainStep;
        return this;
    }
    /**
     * Click on an element
     */
    click(selector) {
        this.addAction({
            action: 'click',
            args: [selector],
        });
        return this;
    }
    /**
     * Type text into an input
     * The input type will be automatically detected during validation if not provided
     */
    type(selector, text, inputType) {
        const args = inputType
            ? [selector, text, inputType] // User specified type
            : [selector, text]; // Will be auto-detected by enricher
        this.addAction({
            action: 'type',
            args: args,
        });
        return this;
    }
    /**
     * Wait for a selector to appear
     */
    waitFor(selector, timeout) {
        this.addAction({
            action: 'waitForSelector',
            args: [selector, { timeout: timeout || 30000 }],
        });
        return this;
    }
    /**
     * Wait for a specific duration
     */
    wait(milliseconds) {
        this.addAction({
            action: 'waitForTimeout',
            args: [milliseconds],
        });
        return this;
    }
    /**
     * Capture a screenshot
     */
    captureScreenshot(name, options) {
        const screenshotArgs = options ? {
            type: options.type || 'png',
            caret: options.caret || 'hide',
            scale: options.scale || 'device',
            timeout: options.timeout || 30000,
            fullPage: options.fullPage !== undefined ? options.fullPage : true,
            animations: options.animations || 'allow',
            ...(options.quality && { quality: options.quality })
        } : {
            type: 'png',
            caret: 'hide',
            scale: 'device',
            timeout: 30000,
            fullPage: true,
            animations: 'allow'
        };
        this.addAction({
            action: 'screenshot',
            name: name,
            args: [screenshotArgs]
        });
        return this;
    }
    /**
     * Scroll the page
     */
    scroll(direction, distance) {
        this.addAction({
            action: 'scroll',
            args: [{ direction, distance }],
        });
        return this;
    }
    /**
     * Set cookies
     */
    setCookies(cookies) {
        if (this.currentStep) {
            this.currentStep.where.cookies = cookies;
        }
        return this;
    }
    /**
     * Set robot mode (normal or bulk)
     */
    mode(mode) {
        this.meta.mode = mode;
        return this;
    }
    /**
     * Set output formats
     */
    format(formats) {
        this.meta.formats = formats;
        return this;
    }
    /**
     * Add a new step to the workflow
     */
    addStep(step) {
        this.currentStep = step;
        this.workflow.unshift(step);
    }
    /**
     * Add an action to the current step
     * Actions are added to the end since what[] executes top-to-bottom
     */
    addAction(action) {
        if (!this.currentStep) {
            // Create a new step if none exists
            this.addStep({
                where: {},
                what: [action],
            });
        }
        else {
            // Use push to add to end - top-to-bottom execution
            this.currentStep.what.push(action);
        }
    }
    /**
     * Get the built workflow (array only)
     */
    getWorkflowArray() {
        return this.workflow;
    }
    /**
     * Get the complete workflow structure with metadata
     */
    getWorkflow() {
        return {
            meta: this.meta,
            workflow: this.workflow
        };
    }
    /**
     * Get the robot metadata
     */
    getMeta() {
        return this.meta;
    }
}
exports.WorkflowBuilder = WorkflowBuilder;
