// maxun-sdk/dist/client/maxun-client.js
"use strict";
/**
 * Maxun API Client
 * Handles all HTTP communication with the Maxun backend API
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = void 0;
const axios_1 = __importDefault(require("axios"));
const http_1 = __importDefault(require("http"));
const https_1 = __importDefault(require("https"));
const types_1 = require("../types");
class Client {
    constructor(config) {
        this.apiKey = config.apiKey;
        this.axios = axios_1.default.create({
            baseURL: config.baseUrl || 'http://localhost:8080/api/sdk',
            headers: {
                'x-api-key': this.apiKey,
                'Content-Type': 'application/json',
                ...(config.teamId ? { 'x-team-id': config.teamId } : {}),
            },
            timeout: 30000,
        });
        // Add response interceptor for error handling
        this.axios.interceptors.response.use((response) => response, (error) => {
            throw this.handleError(error);
        });
    }
    /**
     * Handle API errors and convert to MaxunError
     */
    handleError(error) {
        if (error.response) {
            const data = error.response.data;
            return new types_1.MaxunError(data?.error || data?.message || 'API request failed', error.response.status, data);
        }
        else if (error.request) {
            return new types_1.MaxunError('No response from server', undefined, error.message);
        }
        else {
            return new types_1.MaxunError(error.message);
        }
    }
    /**
     * Get all robots for the authenticated user
     */
    async getRobots() {
        const response = await this.axios.get('/robots');
        return response.data.data || [];
    }
    /**
     * Get a specific robot by ID
     */
    async getRobot(robotId) {
        const response = await this.axios.get(`/robots/${robotId}`);
        if (!response.data.data) {
            throw new types_1.MaxunError(`Robot ${robotId} not found`, 404);
        }
        return response.data.data;
    }
    /**
     * Create a new robot
     */
    async createRobot(workflowFile) {
        // Create a fresh HTTP agent for this request to avoid stale connections
        const httpAgent = new http_1.default.Agent({ keepAlive: false });
        const httpsAgent = new https_1.default.Agent({ keepAlive: false });
        const robotTypeValue = workflowFile.meta?.robotType || workflowFile.meta?.type;
        const payload = {
            ...workflowFile,
            meta: {
                ...workflowFile.meta,
                type: robotTypeValue,
                robotType: robotTypeValue,
            }
        };
        const response = await this.axios.post('/robots', payload, {
            timeout: 120000,
            httpAgent,
            httpsAgent,
        });
        if (!response.data.data) {
            throw new types_1.MaxunError('Failed to create robot');
        }
        return response.data.data;
    }
    /**
     * Update an existing robot
     */
    async updateRobot(robotId, updates) {
        const response = await this.axios.put(`/robots/${robotId}`, updates);
        if (!response.data.data) {
            throw new types_1.MaxunError(`Failed to update robot ${robotId}`);
        }
        return response.data.data;
    }
    /**
     * Delete a robot
     */
    async deleteRobot(robotId) {
        await this.axios.delete(`/robots/${robotId}`);
    }
    /**
     * Execute a robot and get results
     */
    async executeRobot(robotId, options) {
        // Execute and wait for completion (endpoint handles waiting)
        const response = await this.axios.post(`/robots/${robotId}/execute`, {
            params: options?.params,
            webhook: options?.webhook,
        }, {
            timeout: options?.timeout || 300000, // 5 minutes default
        });
        if (!response.data.data) {
            throw new types_1.MaxunError('Failed to execute robot');
        }
        return response.data.data;
    }
    /**
     * Get all runs for a robot
     */
    async getRuns(robotId) {
        const response = await this.axios.get(`/robots/${robotId}/runs`);
        return response.data.data || [];
    }
    /**
     * Get a specific run by ID
     */
    async getRun(robotId, runId) {
        const response = await this.axios.get(`/robots/${robotId}/runs/${runId}`);
        if (!response.data.data) {
            throw new types_1.MaxunError(`Run ${runId} not found`, 404);
        }
        return response.data.data;
    }
    /**
     * Abort a running or queued run
     */
    async abortRun(robotId, runId) {
        await this.axios.post(`/robots/${robotId}/runs/${runId}/abort`);
    }
    /**
     * Schedule a robot for periodic execution
     */
    async scheduleRobot(robotId, schedule) {
        const response = await this.axios.put(`/robots/${robotId}`, { schedule });
        if (!response.data.data) {
            throw new types_1.MaxunError(`Failed to schedule robot ${robotId}`);
        }
        return response.data.data;
    }
    /**
     * Remove schedule from a robot
     */
    async unscheduleRobot(robotId) {
        const response = await this.axios.put(`/robots/${robotId}`, { schedule: null });
        if (!response.data.data) {
            throw new types_1.MaxunError(`Failed to unschedule robot ${robotId}`);
        }
        return response.data.data;
    }
    /**
     * Add a webhook to a robot
     */
    async addWebhook(robotId, webhook) {
        const robot = await this.getRobot(robotId);
        const webhooks = robot.webhooks || [];
        // Add webhook with proper structure
        const newWebhook = {
            id: `webhook_${Date.now()}`,
            url: webhook.url,
            events: webhook.events || ['run.completed', 'run.failed'],
            active: true,
            createdAt: new Date().toISOString(),
            updatedAt: new Date().toISOString(),
        };
        webhooks.push(newWebhook);
        const response = await this.axios.put(`/robots/${robotId}`, {
            webhooks,
        });
        if (!response.data.data) {
            throw new types_1.MaxunError(`Failed to add webhook to robot ${robotId}`);
        }
        return response.data.data;
    }
    /**
     * LLM-based extraction - extract data using natural language prompt
     * URL is optional - if not provided, the server will search for the target website based on the prompt
     */
    async extractWithLLM(options) {
        const response = await this.axios.post('/extract/llm', {
            url: options.url || undefined,
            prompt: options.prompt,
            llmProvider: options.llmProvider,
            llmModel: options.llmModel,
            llmApiKey: options.llmApiKey,
            llmBaseUrl: options.llmBaseUrl,
            robotName: options.robotName,
        }, {
            timeout: 300000,
        });
        if (!response.data.data) {
            throw new types_1.MaxunError('Failed to extract data with LLM');
        }
        return response.data.data;
    }
    /**
     * Create a crawl robot to discover and scrape multiple pages
     */
    async createCrawlRobot(url, options) {
        const response = await this.axios.post('/crawl', {
            url,
            name: options.name,
            crawlConfig: options.crawlConfig,
        });
        if (!response.data.data) {
            throw new types_1.MaxunError('Failed to create crawl robot');
        }
        return response.data.data;
    }
    /**
     * Create a search robot to search and scrape search results
     */
    async createSearchRobot(options) {
        const response = await this.axios.post('/search', {
            name: options.name,
            searchConfig: options.searchConfig,
        });
        if (!response.data.data) {
            throw new types_1.MaxunError('Failed to create search robot');
        }
        return response.data.data;
    }
}
exports.Client = Client;
