// maxun-sdk/dist/robot/robot.js
"use strict";
/**
 * Robot class - represents a saved workflow that can be executed
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Robot = void 0;
class Robot {
    constructor(client, robotData) {
        this.client = client;
        this.robotData = robotData;
    }
    /**
     * Get the robot ID
     */
    get id() {
        return this.robotData.recording_meta.id;
    }
    /**
     * Get the robot name
     */
    get name() {
        return this.robotData.recording_meta.name;
    }
    /**
     * Get the full robot data
     */
    getData() {
        return this.robotData;
    }
    /**
     * Execute the robot
     */
    async run(options) {
        return await this.client.executeRobot(this.id, options);
    }
    /**
     * Get all runs for this robot
     */
    async getRuns() {
        return await this.client.getRuns(this.id);
    }
    /**
     * Get a specific run
     */
    async getRun(runId) {
        return await this.client.getRun(this.id, runId);
    }
    /**
     * Get the latest run
     */
    async getLatestRun() {
        const runs = await this.getRuns();
        if (runs.length === 0)
            return null;
        // Sort by startedAt descending
        runs.sort((a, b) => new Date(b.startedAt).getTime() - new Date(a.startedAt).getTime());
        return runs[0];
    }
    /**
     * Abort a running or queued run
     */
    async abort(runId) {
        await this.client.abortRun(this.id, runId);
    }
    /**
     * Schedule the robot for periodic execution
     */
    async schedule(config) {
        const updated = await this.client.scheduleRobot(this.id, config);
        this.robotData = updated;
    }
    /**
     * Remove the schedule
     */
    async unschedule() {
        const updated = await this.client.unscheduleRobot(this.id);
        this.robotData = updated;
    }
    /**
     * Add a webhook
     */
    async addWebhook(webhook) {
        const updated = await this.client.addWebhook(this.id, webhook);
        this.robotData = updated;
    }
    /**
     * Update the robot's workflow or metadata
     */
    async update(updates) {
        const updated = await this.client.updateRobot(this.id, updates);
        this.robotData = updated;
    }
    /**
     * Get all webhooks for this robot
     */
    getWebhooks() {
        return this.robotData.webhooks || null;
    }
    /**
     * Remove all webhooks
     */
    async removeWebhooks() {
        const updated = await this.client.updateRobot(this.id, {
            webhooks: null,
        });
        this.robotData = updated;
    }
    /**
     * Get schedule configuration
     */
    getSchedule() {
        return this.robotData.schedule || null;
    }
    /**
     * Delete the robot
     */
    async delete() {
        await this.client.deleteRobot(this.id);
    }
    /**
     * Refresh robot data from server
     */
    async refresh() {
        this.robotData = await this.client.getRobot(this.id);
    }
}
exports.Robot = Robot;
