// maxun-sdk/dist/llm/providers/anthropic.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnthropicProvider = void 0;
const sdk_1 = __importDefault(require("@anthropic-ai/sdk"));
const base_1 = require("./base");
/**
 * Anthropic provider for Claude models
 */
class AnthropicProvider extends base_1.BaseLLMProvider {
    constructor(config) {
        super(config);
        this.client = new sdk_1.default({
            apiKey: config.apiKey || process.env.ANTHROPIC_API_KEY
        });
    }
    async chat(messages) {
        try {
            // Anthropic requires system message separately
            const systemMessage = messages.find(m => m.role === 'system');
            const userMessages = messages.filter(m => m.role !== 'system');
            const response = await this.client.messages.create({
                model: this.config.model || 'claude-3-5-sonnet-20241022',
                max_tokens: this.config.maxTokens || 4096,
                temperature: this.config.temperature || 0.7,
                system: systemMessage?.content,
                messages: userMessages.map(msg => ({
                    role: msg.role,
                    content: msg.content
                }))
            });
            const textContent = response.content.find((c) => c.type === 'text');
            return {
                content: textContent?.type === 'text' ? textContent.text : '',
                usage: {
                    promptTokens: response.usage.input_tokens,
                    completionTokens: response.usage.output_tokens,
                    totalTokens: response.usage.input_tokens + response.usage.output_tokens
                }
            };
        }
        catch (error) {
            throw new Error(`Anthropic API error: ${error.message || error}`);
        }
    }
    validateConfig() {
        if (!this.config.apiKey && !process.env.ANTHROPIC_API_KEY) {
            throw new Error('Anthropic API key required. Set ANTHROPIC_API_KEY environment variable or provide apiKey in config.');
        }
        if (!this.config.model) {
            console.warn('Anthropic model not specified, using default: claude-3-5-sonnet-20241022');
        }
    }
    getProviderName() {
        return 'Anthropic';
    }
}
exports.AnthropicProvider = AnthropicProvider;
