// maxun-sdk/dist/llm/providers/ollama.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OllamaProvider = void 0;
const ollama_1 = require("ollama");
const base_1 = require("./base");
/**
 * Ollama provider for local LLM inference
 * Supports Llama, Mistral, and other open-source models
 */
class OllamaProvider extends base_1.BaseLLMProvider {
    constructor(config) {
        super(config);
        this.client = new ollama_1.Ollama({
            host: config.baseUrl || process.env.OLLAMA_HOST || 'http://localhost:11434'
        });
    }
    async chat(messages) {
        try {
            const response = await this.client.chat({
                model: this.config.model || 'llama3.1',
                messages: messages.map(msg => ({
                    role: msg.role,
                    content: msg.content
                })),
                options: {
                    temperature: this.config.temperature || 0.7,
                    num_predict: this.config.maxTokens || 2048
                }
            });
            return {
                content: response.message.content,
                usage: {
                    promptTokens: response.prompt_eval_count || 0,
                    completionTokens: response.eval_count || 0,
                    totalTokens: (response.prompt_eval_count || 0) + (response.eval_count || 0)
                }
            };
        }
        catch (error) {
            throw new Error(`Ollama API error: ${error.message || error}`);
        }
    }
    validateConfig() {
        if (!this.config.baseUrl && !process.env.OLLAMA_HOST) {
            console.warn('Ollama baseUrl not provided, using default: http://localhost:11434');
        }
        if (!this.config.model) {
            console.warn('Ollama model not specified, using default: llama3.1');
        }
    }
    getProviderName() {
        return 'Ollama';
    }
}
exports.OllamaProvider = OllamaProvider;
