// maxun-sdk/dist/llm/providers/openai.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenAIProvider = void 0;
const openai_1 = __importDefault(require("openai"));
const base_1 = require("./base");
/**
 * OpenAI provider for GPT models
 */
class OpenAIProvider extends base_1.BaseLLMProvider {
    constructor(config) {
        super(config);
        this.client = new openai_1.default({
            apiKey: config.apiKey || process.env.OPENAI_API_KEY,
            baseURL: config.baseUrl // Allows custom endpoints
        });
    }
    async chat(messages) {
        try {
            const response = await this.client.chat.completions.create({
                model: this.config.model || 'gpt-4o-mini',
                messages: messages.map(msg => ({
                    role: msg.role,
                    content: msg.content
                })),
                temperature: this.config.temperature || 0.7,
                max_tokens: this.config.maxTokens || 4096
            });
            const choice = response.choices[0];
            return {
                content: choice.message.content || '',
                usage: {
                    promptTokens: response.usage?.prompt_tokens || 0,
                    completionTokens: response.usage?.completion_tokens || 0,
                    totalTokens: response.usage?.total_tokens || 0
                }
            };
        }
        catch (error) {
            throw new Error(`OpenAI API error: ${error.message || error}`);
        }
    }
    validateConfig() {
        if (!this.config.apiKey && !process.env.OPENAI_API_KEY) {
            throw new Error('OpenAI API key required. Set OPENAI_API_KEY environment variable or provide apiKey in config.');
        }
        if (!this.config.model) {
            console.warn('OpenAI model not specified, using default: gpt-4o-mini');
        }
    }
    getProviderName() {
        return 'OpenAI';
    }
}
exports.OpenAIProvider = OpenAIProvider;
