// maxun-sdk/dist/llm/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenAIProvider = exports.AnthropicProvider = exports.OllamaProvider = exports.BaseLLMProvider = void 0;
exports.createLLMProvider = createLLMProvider;
const ollama_1 = require("./providers/ollama");
const anthropic_1 = require("./providers/anthropic");
const openai_1 = require("./providers/openai");
/**
 * Factory function to create LLM provider instances
 */
function createLLMProvider(config) {
    switch (config.provider) {
        case 'ollama':
            return new ollama_1.OllamaProvider(config);
        case 'anthropic':
            return new anthropic_1.AnthropicProvider(config);
        case 'openai':
            return new openai_1.OpenAIProvider(config);
        default:
            throw new Error(`Unsupported LLM provider: ${config.provider}`);
    }
}
// Re-export types and base class
var base_1 = require("./providers/base");
Object.defineProperty(exports, "BaseLLMProvider", { enumerable: true, get: function () { return base_1.BaseLLMProvider; } });
var ollama_2 = require("./providers/ollama");
Object.defineProperty(exports, "OllamaProvider", { enumerable: true, get: function () { return ollama_2.OllamaProvider; } });
var anthropic_2 = require("./providers/anthropic");
Object.defineProperty(exports, "AnthropicProvider", { enumerable: true, get: function () { return anthropic_2.AnthropicProvider; } });
var openai_2 = require("./providers/openai");
Object.defineProperty(exports, "OpenAIProvider", { enumerable: true, get: function () { return openai_2.OpenAIProvider; } });
