// maxun-sdk/dist/extract.js
"use strict";
/**
 * Extract - Main class for the Extract SDK
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Extract = void 0;
const maxun_client_1 = require("./client/maxun-client");
const extract_builder_1 = require("./builders/extract-builder");
const robot_1 = require("./robot/robot");
class Extract {
    constructor(config) {
        this.client = new maxun_client_1.Client(config);
    }
    /**
     * Create a new extraction workflow
     */
    create(name) {
        const builder = new extract_builder_1.ExtractBuilder(name);
        builder.setExtractor(this);
        return builder;
    }
    /**
     * Build and save the robot to Maxun
     */
    async build(builder) {
        const workflow = builder.getWorkflowArray();
        const meta = builder.getMeta();
        // Generate a unique ID for the robot
        const robotId = `robot_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
        meta.id = robotId;
        const workflowFile = {
            meta: meta,
            workflow,
        };
        // Create the robot
        const robot = await this.client.createRobot(workflowFile);
        return new robot_1.Robot(this.client, robot);
    }
    /**
     * Get all extract robots
     */
    async getRobots() {
        const robots = await this.client.getRobots();
        const extractRobots = robots.filter((r) => r.recording_meta.robotType === 'extract');
        return extractRobots.map((r) => new robot_1.Robot(this.client, r));
    }
    /**
     * Get a specific robot by ID
     */
    async getRobot(robotId) {
        const robot = await this.client.getRobot(robotId);
        return new robot_1.Robot(this.client, robot);
    }
    /**
     * Delete a robot
     */
    async deleteRobot(robotId) {
        await this.client.deleteRobot(robotId);
    }
    /**
     * LLM-based extraction - create a robot using natural language prompt
     * The robot is saved and can be executed anytime by the user
     *
     * @param options - Extraction options
     * @param options.url - (Optional) The URL to extract data from. If not provided, the system will automatically search for the target website based on the prompt.
     * @param options.prompt - Natural language prompt describing what to extract
     * @param options.llmProvider - LLM provider to use: 'anthropic', 'openai', or 'ollama' (default: 'ollama')
     * @param options.llmModel - Model name (default: 'llama3.2-vision' for ollama, 'claude-3-5-sonnet-20241022' for anthropic, 'gpt-4-vision-preview' for openai)
     * @param options.llmApiKey - API key for the LLM provider (not needed for ollama)
     * @param options.llmBaseUrl - Base URL for the LLM provider (default: 'http://localhost:11434' for ollama)
     * @param options.robotName - Optional custom name for the robot
     * @returns Robot instance that can be executed
     */
    async extract(options) {
        const robotData = await this.client.extractWithLLM(options);
        const robot = await this.client.getRobot(robotData.robotId);
        return new robot_1.Robot(this.client, robot);
    }
}
exports.Extract = Extract;
