// maxun-sdk/dist/index.js
"use strict";
/**
 * Maxun SDK - Unified package for web automation and data extraction
 *
 * @packageDocumentation
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = exports.Robot = exports.Search = exports.Crawl = exports.Scrape = exports.Extract = void 0;
// Main classes
var extract_1 = require("./extract");
Object.defineProperty(exports, "Extract", { enumerable: true, get: function () { return extract_1.Extract; } });
var scrape_1 = require("./scrape");
Object.defineProperty(exports, "Scrape", { enumerable: true, get: function () { return scrape_1.Scrape; } });
var crawl_1 = require("./crawl");
Object.defineProperty(exports, "Crawl", { enumerable: true, get: function () { return crawl_1.Crawl; } });
var search_1 = require("./search");
Object.defineProperty(exports, "Search", { enumerable: true, get: function () { return search_1.Search; } });
var robot_1 = require("./robot/robot");
Object.defineProperty(exports, "Robot", { enumerable: true, get: function () { return robot_1.Robot; } });
var maxun_client_1 = require("./client/maxun-client");
Object.defineProperty(exports, "Client", { enumerable: true, get: function () { return maxun_client_1.Client; } });
// All types
__exportStar(require("./types"), exports);
