// maxun-sdk/dist/scrape.js
"use strict";
/**
 * Scrape - Main class for the Scrape SDK
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Scrape = void 0;
const maxun_client_1 = require("./client/maxun-client");
const robot_1 = require("./robot/robot");
class Scrape {
    constructor(config) {
        this.client = new maxun_client_1.Client(config);
    }
    /**
     * Create a new scraping robot
     * @param name - Name of the scraping robot
     * @param url - URL to scrape
     * @param options - Optional scraping options (formats)
     * @returns Promise<Robot>
     */
    async create(name, url, options) {
        if (!url) {
            throw new Error('URL is required');
        }
        const workflowFile = {
            meta: {
                name,
                id: `robot_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
                robotType: 'scrape',
                url,
                formats: options?.formats || ['markdown'],
            },
            workflow: [],
        };
        const robot = await this.client.createRobot(workflowFile);
        return new robot_1.Robot(this.client, robot);
    }
}
exports.Scrape = Scrape;
